# Abbreviation Disambiguation in Medical Texts - Data Modeling

This Notebook is in continuation of the notebook- 'Step 2- Data Preprocessing' and lists down:

1. Modeling Preprocessed data using: GridSearchCV on Logistic Regression, SVM and XG Boost.
2. Testing the models using Test set.
3. Comparing the models and identifying the Next Steps

## Step# 1: Loading Dataset

In [2]:
#Importing the Required Python Packages
import shutil
import string
import numpy as np
import pandas as pd
import pickle
from tqdm import tqdm
import ast
from sklearn import utils
from gensim.models import Doc2Vec
from gensim.models.doc2vec import TaggedDocument
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import f1_score, accuracy_score, classification_report
import matplotlib.pyplot as plt
import xgboost as xgb
from xgboost.sklearn import XGBClassifier
pd.set_option('display.max_colwidth', None)

In [3]:
# Lets load the train dataset.
train = pd.read_csv('Train/train_final.csv')
train.head(3)

LABEL ABV  \
0      anticoagulation  AC   
1      anticoagulation  AC   
2  typical development  TD   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       TOKEN  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                    ['yearold', 'girl', 'presented', 'unilateral', 'c6', 'nerve', 'palsy', 'papilledema', 't3', 'severe', 'ha', 'magnetic', 'resonance', 'imaging', 'showed', 'thrombosis', 'superior', 'sagittal', 'sinus', 'pt', 'right', 'transverse', 'sinus', 'attributed', 'oral', 'contraceptive', 'use', 't3', 'coagulation', 'workup', 'negative', 'svr', 'anticoagulation', 'caused', 'hemorrhagic', 'papillopathy', 'eyes', 'raising', 'question', 'ac', 'discontinued', 'diagnostic', 'management', 'issues', 'cbf', 'venous', 'sinus', 'thrombosis', 'secondary', 'intracranial', 'hypertension', 'discussed']  
1                                                                                                                                                               ['mvr', 'procedure', 'choice', 'treat', 'mitral', 'valve', 'dysfunction', 'advantages', 'mvr', 'mitral', 'vr', 'include', 'improved', 'lts', 'better', 'preservation', 'left', 'vvi', 'cf', 'greater', 'freedom', 'endocarditis', 'thromboembolism', 'anticoagulantrelated', 'hemorrhage', 'feasibility', 'durability', 'mvr', 'depend', 'etiology', 'mitral', 'valve', 'dysfunction', 'degenerative', 'icm', 'mitral', 'valve', 'diseases', 'valve', 'repair', 'possible', 'cases', 'freedom', 'reoperation', 'high', 'contrast', 'rheumatic', 'valves', 'amenable', 'repair', 'durability', 'limited', 'isolated', 'valve', 'operations', 'performed', 'minimally', 'invasive', 'fashion', 'partial', 'upper', 'sternotomy', 'patients', 'atrial', 'fibrillation', 'mvd', 'maze', 'procedure', 'pulmonary', 'vein', 'isolation', 'performed', 'eliminating', 'atrial', 'fibrillation', 'need', 'lt', 'ac']  
2  ['explored', 'classification', 'accuracy', 'english', 'spanish', 'versions', 'experimental', 'semantic', 'language', 'measure', 'functional', 'monolingualbilingual', 'children', 'language', 'impairment', 'total', 'children', 'participated', 'including', 'balanced', 'bilinguals', 'language', 'impairment', 'td', 'monolingual', 'spanish', 'language', 'impairment', 'typical', 'od', 'monolingual', 'english', 'language', 'impairment', 'typical', 'od', 'children', 'years', 'cut', 'points', 'derived', 'functionally', 'monolingual', 'children', 'applied', 'bilinguals', 'assess', 'predictive', 'accuracy', 'english', 'spanish', 'semantics', 'correct', 'classification', 'english', 'monolinguals', 'spanish', 'monolinguals', 'discriminant', 'analysis', 'yielded', 'correct', 'clas

In [4]:
# Lets load validation and test datasets as well
valid = pd.read_csv('Validation/valid_final.csv')
test = pd.read_csv('Test/test_final.csv')

In [5]:
valid.head(3)

LABEL ABV  \
0           benzophenone  BP   
1                  adult  AD   
2  primary visual cortex  V1   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              TOKEN  
0                                                                                                                                                                                                                                                     ['bp', 'fema', 'cas', 'po', 'diet', 'rats', 'target', 'dls', 'mgkg', 'body', 'weightday', 'days', 'mgkgday', 'days', 'body', 'weights', 'food', 'consumption', 'measured', 'weekly', 'haematology', 'clinical', 'chemistry', 'urinalysis', 'values', 'obtained', 'wk', 'end', 'study', 'gross', 'microscopic', 'pathological', 'examinations', 'conducted', 'organ', 'weights', 'recorded', 'treatmentrelated', 'changes', 'occurred', 'ea', 'count', 'haemoglobin', 'haematocrit', 'bilirubin', 'total', 'protein', 'albumin', 'mid', 'highdose', 'c2', 'changes', 'occur', 'groups', 'sexes', 'indications', 'increased', 'absolute', 'relative', 'cl', 'kidney', 'weights', 'mid', 'highdose', 'groups', 'statistically', 'consistent', 'absolute', 'kidney', 'weights', 'histopathology', 'cl', 'mid', 'highdose', 'cg', 'showed', 'hepatocellular', 'enlargement', 'associated', 'clumping', 'cytoplasmic', 'basophilic', 'material', 'central', 'vein', 'nel', 'demonstrated', 'mgkgday', 'days', 'administration', 'equivalent', 'ni', 'mgday', 'kg', 'human', 'basis', 'calculated', 'possible', 'average', 'daily', 'intake', 'mgday', 'safety', 'factor', 'greater', 'demonstrated', 'safety', 'factor', 'based', 'realistic', 'capita', 'consumption', 'microgramday', 'approximately', 'million']  
1  ['investigated', 'expression', 'xist', 'gene', 'mouse', 'female', 'ad', 'kidney', 'embryos', 'embryonic', 'stem', 'es', 'cells', 'undergoing', 'vitro', 'differentiation', 'ebs', 'quantitative', 'rtpcr', 'single', 'nucleotide', 'primer', 'extension', 'snupe', 'ca', 'found', 'xist', 'rna', 'ad', 'kidney', 'mouse', 'strains', 'approximately', 'transcripts', 'cell', 'modest', 'differences', 'strains', 'carrying', 'different', 'xce', 'alleles', 'female', 'embryos', 'days', 'pc', 'number', 'xist', 'transcripts', 'cell', 'isogenic', 'ad', 'tissue', 'quantitative', 'oligonucleotide', 'hybridization', 'assays', 't3', 'rtpcr', 'investigated', 'xist', 'expression', 'es', 'lines', 'heterozygous', 'pgk', 'xist', 'loci', 'found', 'xx', 'es', 'lines', 'xist', 'rna', 'levels', 'increased', 'embryoid', 'body', 'formation', 'levels', 'seen', 'found', 'adult', 'female', 'ki

In [6]:
test.head(3)

LABEL ABV  \
0         micronucleus  MN   
1              bipolar  BP   
2  genetic programming  GP   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             TOKEN  
0                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          ['work', 'tested', 'hypothesis', 'content', 'spontaneous', 'micronuclei', 'lymphocytes', 'apparently', 'healthy', 'normal', 'human', 'subject', 'exhibited', 'unusually', 'high', 'mn', 'frequency', 'nonrandom', 'dna', 'probes', 'fluorescent', 'insitu', 'hybridization', 'fish', 'beginning', 'probe', 'generated', 'subjects', 'micronuclei', 'micronuclei', 'obtained', 'peripheral', 'bl', 'microdissection', 'subjected', 'rapd', 'rapdpcr', 'unique', 'pcr', 'product', 'isolate', 'cosmid', 'clone', 'human', 'genomic', 'library', 'clone', 'hybridized', 'chromosome', 'subsequently', 'commercial', 'probes', 'included', 'fish', 'analyses', 'micronuclei', 'subject', 'age', 'sexmatched', 'controls', 'significant', 'differences', 'found', 'subject', 'controls', 'percentages', 'micronuclei', 'hybridizing', 'centromere', 'probe', 'x', 'chromosome', 'painting', 'probe', 'chromosome', 'subject', 'highly', 'significant', 'increase', 'p', 'chromosome', 'micronuclei', 'level', 'expected', 'present', 'chance', 'characterization', 'micronuclei', 'promising', 'tool', 'studies', 'mechanisms', 'inherited', 'induced', 'cin', 'strength', 'strategy', 'employed', 't0', 'characterizing', 'chromosomes', 'present', 'micronuclei', 'work', 't3', 'observation', 'chromosomal', 'instability', 'foundation', 't0', 'mechanism', 'underlying', 'observation']  
1  ['onethird', 'lithiumtreated', 'bp', 'patients', 'excellent', 'lithium', 'nr', 'lithium', 'monotherapy', 't

### Lets keep only relevant records in Valid and test set.

In [7]:
    abbrev = list(train['ABV'].unique())
    valid = valid[valid['ABV'].isin(abbrev)]
    test = test[test['ABV'].isin(abbrev)]
    labels = list(train['LABEL'].unique())
    valid = valid[valid['LABEL'].isin(labels)]
    test = test[test['LABEL'].isin(labels)]


### Lets drop 'ABV' column

In [8]:
train.drop(columns='ABV', inplace = True)
valid.drop(columns='ABV', inplace = True)
test.drop(columns='ABV', inplace = True)

In [9]:
# Convert TOKEN column from string to list
train['TOKEN'] = train['TOKEN'].apply(lambda x: ast.literal_eval(x))
valid['TOKEN'] = valid['TOKEN'].apply(lambda x: ast.literal_eval(x))
test['TOKEN'] = test['TOKEN'].apply(lambda x: ast.literal_eval(x))

### Lets tag every Token List with its Label

In [10]:
train_tagged = train.apply(lambda x: TaggedDocument(words = x['TOKEN'], tags = [x['LABEL']]), axis=1)
valid_tagged = valid.apply(lambda x: TaggedDocument(words = x['TOKEN'], tags = [x['LABEL']]), axis=1)
test_tagged = test.apply(lambda x: TaggedDocument(words = x['TOKEN'], tags = [x['LABEL']]), axis=1)

In [11]:
train_tagged.values[:5]

array([TaggedDocument(words=['yearold', 'girl', 'presented', 'unilateral', 'c6', 'nerve', 'palsy', 'papilledema', 't3', 'severe', 'ha', 'magnetic', 'resonance', 'imaging', 'showed', 'thrombosis', 'superior', 'sagittal', 'sinus', 'pt', 'right', 'transverse', 'sinus', 'attributed', 'oral', 'contraceptive', 'use', 't3', 'coagulation', 'workup', 'negative', 'svr', 'anticoagulation', 'caused', 'hemorrhagic', 'papillopathy', 'eyes', 'raising', 'question', 'ac', 'discontinued', 'diagnostic', 'management', 'issues', 'cbf', 'venous', 'sinus', 'thrombosis', 'secondary', 'intracranial', 'hypertension', 'discussed'], tags=['anticoagulation']),
       TaggedDocument(words=['mvr', 'procedure', 'choice', 'treat', 'mitral', 'valve', 'dysfunction', 'advantages', 'mvr', 'mitral', 'vr', 'include', 'improved', 'lts', 'better', 'preservation', 'left', 'vvi', 'cf', 'greater', 'freedom', 'endocarditis', 'thromboembolism', 'anticoagulantrelated', 'hemorrhage', 'feasibility', 'durability', 'mvr', 'depend', 'et

## Step# 2: Apply Doc2vec vectorizer on the Dataset

In [12]:
vectorize = Doc2Vec(dm=0, vector_size=100, min_count=2, window = 2)
vectorize.build_vocab(train_tagged.values)

In [13]:
vectorize.train(train_tagged.values, total_examples=len(train_tagged.values), epochs=30)

### Building the Final Vector Feature Classifier

In [14]:
def vec_for_learning(model, tagged_docs):
    sents = tagged_docs.values
    targets, regressors = zip(*[(doc.tags[0], model.infer_vector(doc.words, steps=30)) for doc in sents])
    return targets, regressors

In [15]:
y_train, X_train = vec_for_learning(vectorize, train_tagged)

## Model# 1: Logistic Classifier

### Lets perform a Grid Search to get the best possible combination of Hyperparameters for Logistic Regression Model

In [41]:
param_grid = {'C':[0.001, 0.01, 0.1, 1, 10, 100]}
grid_model = GridSearchCV(LogisticRegression(n_jobs=-1), param_grid)

In [42]:
grid_model.fit(X_train, y_train)

GridSearchCV(estimator=LogisticRegression(n_jobs=-1),
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100]})

In [43]:
### Best parameters for the Grid Search
grid_model.best_params_

{'C': 1}

In [44]:
### Accuracy Score
grid_model.best_score_

0.9805815760310533

### Apply the best parameters to Logistic Regression and train the model.

In [45]:
logreg = LogisticRegression(n_jobs=-1, C=1)
logreg.fit(X_train, y_train)

LogisticRegression(C=1, n_jobs=-1)

In [46]:
### Apply the above Model on Validation Set
y_valid, X_valid = vec_for_learning(vectorize, valid_tagged)
y_pred_valid = logreg.predict(X_valid)

In [47]:
print('Validation Accuracy:', accuracy_score(y_valid, y_pred_valid))
print('Validation F1-Score:', f1_score(y_valid, y_pred_valid, average='weighted'))

Validation Accuracy: 0.6913
Validation F1-Score: 0.6848102160464058


### As per the above analysis of validation set, it can be seen that the Logistic Classification model gives
1. F1- Score of: 0.68
2. Accuracy of: 69%

Hence, lets apply this model to our Test set and check its performance metrics.

In [48]:
### Apply the above Model on Test Set
y_test, X_test = vec_for_learning(vectorize, test_tagged)
y_pred_test = logreg.predict(X_test)

### Lets calculate some Performance Metrics on the Test predictions.

In [68]:
accuracy = accuracy_score(y_test, y_pred_test)
f1_scr = f1_score(y_test, y_pred_test, average='weighted')
print('Test Accuracy:', accuracy)
print('Test F1-Score:', f1_scr)

Test Accuracy: 0.6962
Test F1-Score: 0.6906125218543699


In [74]:
y_unique = list(set(y_test))
cr = classification_report(y_test, y_pred_test, target_names=y_unique)
print(cr)

                                                   precision    recall  f1-score   support

                                habitual abortion       0.83      1.00      0.91        15
                                 hemolytic anemia       0.79      0.91      0.85        45
                                   thermodilution       0.76      0.60      0.67        48
                               activated charcoal       0.84      0.88      0.86        48
                                  edmonstonzagreb       0.42      0.30      0.35        54
                       metabolism of benzoapyrene       0.95      1.00      0.98        21
                                  buffer capacity       0.79      0.93      0.85        41
                                       decay rate       1.00      0.78      0.88         9
                          hypersensitive reaction       0.96      0.86      0.91        29
                      live attenuated coldadapted       0.87      0.85      0.86        5

### Thus, from the above Report it can be seen:
1. Average Precision: 0.69
2. Average Recall: 0.70

## Model# 2: SVM

### Lets perform a Grid Search to get the best possible combination of Hyperparameters for SVM's

In [15]:
param_grid = {'C': [0.001, 0.01, 0.1, 1, 10, 100],'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 'kernel': ['rbf']}
grid_svm = GridSearchCV(SVC(), param_grid)

In [16]:
grid_svm.fit(X_train, y_train)

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100],
                         'gamma': [1, 0.1, 0.01, 0.001, 0.0001],
                         'kernel': ['rbf']})

In [17]:
### Best parameters for the Grid Search
grid_svm.best_params_

{'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}

In [18]:
### Accuracy Score
grid_svm.best_score_

0.9725555822642471

### Apply the best parameters to SVC and train the model

In [19]:
svcModel = SVC(C=10, gamma=0.01, kernel='rbf')
svcModel.fit(X_train, y_train)

SVC(C=10, gamma=0.01)

In [20]:
### Apply the above Model on Validation Set
y_valid, X_valid = vec_for_learning(vectorize, valid_tagged)
y_pred_valid = svcModel.predict(X_valid)

In [21]:
print('SVM Validation Accuracy:', accuracy_score(y_valid, y_pred_valid))
print('SVM Validation F1-Score:', f1_score(y_valid, y_pred_valid, average='weighted'))

SVM Validation Accuracy: 0.7067
SVM Validation F1-Score: 0.7028056336697289


### As per the above analysis of validation set, it can be seen that the SVC Classification model gives
1. F1- Score of: 0.71
2. Accuracy of: 70%

Hence, lets apply this model to our Test set and check its performance metrics.

In [22]:
### Apply the above Model on Test Set
y_test, X_test = vec_for_learning(vectorize, test_tagged)
y_pred_test = svcModel.predict(X_test)

### Lets calculate some Performance Metrics on the Test predictions.

In [23]:
accuracy = accuracy_score(y_test, y_pred_test)
f1_scr = f1_score(y_test, y_pred_test, average='weighted')
print('SVM Test Accuracy:', accuracy)
print('SVM Test F1-Score:', f1_scr)

SVM Test Accuracy: 0.7055
SVM Test F1-Score: 0.7009447611264659


In [24]:
y_unique = list(set(y_test))
cr = classification_report(y_test, y_pred_test, target_names=y_unique)
print(cr)

                                                   precision    recall  f1-score   support

                             ischaemiareperfusion       1.00      1.00      1.00        15
                                       fulllength       0.84      0.96      0.90        45
                                   hepatic artery       0.71      0.67      0.69        48
                                       acebutolol       0.87      0.83      0.85        48
                           attenuation correction       0.39      0.41      0.40        54
                                 human fibroblast       0.88      1.00      0.93        21
                               thiamine deficient       0.81      0.85      0.83        41
                                  contrast agents       0.89      0.89      0.89         9
                               firefly luciferase       0.93      0.90      0.91        29
                                arm circumference       0.85      0.85      0.85        5

### Thus, from the above Report it can be seen:
1. Average Precision: 0.71
2. Average Recall: 0.71

## Model# 3: XGBoost

### Lets create a parameter grid for XGBoost Model

In [16]:
param_grid = {'n_estimators':[100, 500, 1000], 'max_depth':[5, 6, 7], 'min_child_weight': [3, 5, 8]}

In [17]:
unique = list(set(y_train))
X_train = pd.DataFrame(X_train)
y_train = np.asarray(y_train)

In [33]:
XGBgrid = GridSearchCV(XGBClassifier(learning_rate= 0.1, gamma= 0, objective= 'multi:softmax', num_classes= len(unique), seed= 27), param_grid)

In [35]:
XGBgrid.fit(X_train, y_train)

C:\Users\shubh\Anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[20:15:08] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:541: 
Parameters: { num_classes } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not used but slip through this
  verification. Please open an issue if you find above cases.


[20:18:28] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.
[00:20:33] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:541: 
Parameters: { num_classes } might not be used.

  This may not be accurate due to some parameters are only used in language bindings but
  passed down to XGBoost core.  Or some parameters are not us

In [ ]:
### Best parameters for the Grid Search
XGBgrid.best_params_

In [ ]:
### Accuracy Score
grid_svm.best_score_

### Train a XGBoost Classifier

In [18]:
XGBModel = XGBClassifier(
 learning_rate =0.1,
 n_estimators=1000,
 max_depth=7,
 min_child_weight=4,
 gamma=0,
 objective= 'multi:softmax',
 seed=27)   
XGBModel.fit(X_train, y_train)

C:\Users\shubh\Anaconda3\lib\site-packages\xgboost\sklearn.py:888: UserWarning: The use of label encoder in XGBClassifier is deprecated and will be removed in a future release. To remove this warning, do the following: 1) Pass option use_label_encoder=False when constructing XGBClassifier object; and 2) Encode your labels (y) as integers starting with 0, i.e. 0, 1, 2, ..., [num_class - 1].
  warnings.warn(label_encoder_deprecation_msg, UserWarning)


[10:57:22] WARNING: C:/Users/Administrator/workspace/xgboost-win64_release_1.3.0/src/learner.cc:1061: Starting in XGBoost 1.3.0, the default evaluation metric used with the objective 'multi:softprob' was changed from 'merror' to 'mlogloss'. Explicitly set eval_metric if you'd like to restore the old behavior.


XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
              importance_type='gain', interaction_constraints='',
              learning_rate=0.1, max_delta_step=0, max_depth=7,
              min_child_weight=4, missing=nan, monotone_constraints='()',
              n_estimators=1000, n_jobs=4, num_parallel_tree=1,
              objective='multi:softprob', random_state=27, reg_alpha=0,
              reg_lambda=1, scale_pos_weight=None, seed=27, subsample=1,
              tree_method='exact', validate_parameters=1, verbosity=None)

In [19]:
### Apply the above Model on Validation Set
y_valid, X_valid = vec_for_learning(vectorize, valid_tagged)
X_valid = pd.DataFrame(X_valid)
y_valid = np.asarray(y_valid)
y_pred_valid = XGBModel.predict(X_valid)

In [20]:
print('XGBoost Validation Accuracy:', accuracy_score(y_valid, y_pred_valid))
print('XGBoost Validation F1-Score:', f1_score(y_valid, y_pred_valid, average='weighted'))

XGBoost Validation Accuracy: 0.5918
XGBoost Validation F1-Score: 0.5861134966489252


### As per the above analysis of validation set, it can be seen that the XGBoost Classification model gives
1. F1- Score of: 0.58
2. Accuracy of: 59.1%

Hence, lets apply this model to our Test set and check its performance metrics.

In [21]:
### Apply the above Model on Test Set
y_test, X_test = vec_for_learning(vectorize, test_tagged)
X_test = pd.DataFrame(X_test)
y_test = np.asarray(y_test)
y_pred_test = XGBModel.predict(X_test)

### Lets calculate some Performance Metrics on the Test predictions.

In [22]:
accuracy = accuracy_score(y_test, y_pred_test)
f1_scr = f1_score(y_test, y_pred_test, average='weighted')
print('XGBoost Test Accuracy:', accuracy)
print('XGBoost Test F1-Score:', f1_scr)

XGBoost Test Accuracy: 0.5891
XGBoost Test F1-Score: 0.5832797485214104


In [23]:
y_unique = list(set(y_test))
cr = classification_report(y_test, y_pred_test, target_names=y_unique)
print(cr)

C:\Users\shubh\Anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


                                                   precision    recall  f1-score   support

                              anterior commissure       0.91      0.67      0.77        15
                    anterodorsal thalamic nucleus       0.68      0.71      0.70        45
                              atypical depression       0.47      0.46      0.46        48
                           artificial circulation       0.80      0.83      0.82        48
                                   striate cortex       0.32      0.26      0.29        54
                            platelet glycoprotein       0.95      0.86      0.90        21
                                 cold agglutinins       0.67      0.76      0.71        41
                                haemagglutination       0.50      0.44      0.47         9
                                            adult       0.80      0.83      0.81        29
                              articular cartilage       0.68      0.81      0.74        5

### Thus, from the above Report it can be seen:
1. Average Precision: 0.59
2. Average Recall: 0.59

# Next Steps

## We can see that a basic Logistic Classification implementation gives 70% Accurate results hence, for next steps we can:
1. Try tuning the Doc2Vec vectorizer's Hyperparameters.
2. Try some other Classification Algorithms like SVN, Random Forrest and compare results.
3. Present model has been trained to disambiguate 20 'Medical Abbreviations' but this same model can be generalized to be used in other fields as well. Some including Scientific Researches and Internet Slags.